File: wrappers/go/ccl/crypto_test.go
package ccl

import (
	"crypto/ed25519"
	"encoding/hex"
	"strings"
	"testing"
)

func TestCryptoGenerate12WordMnemonic(t *testing.T) {
	mnemonic, err := bridge.Crypto.GenerateMnemonic(12)
	if err != nil {
		t.Fatalf("Crypto.GenerateMnemonic(12) failed: %v", err)
	}
	if words := strings.Fields(mnemonic); len(words) != 12 {
		t.Errorf("expected 12 words, got %d", len(words))
	}
	if !bridge.Crypto.ValidateMnemonic(mnemonic) {
		t.Error("generated 12-word mnemonic should validate")
	}
}

// A real Ed25519 vector: the bridge signs a 32-byte seed key and the result must match Go's stdlib
// (Ed25519 is deterministic per RFC 8032), and the signature must verify against the matching public
// key. This makes the sign/verify assertions concrete rather than length-only.
func TestCryptoSignVerifyVector(t *testing.T) {
	seed := make([]byte, ed25519.SeedSize)
	for i := range seed {
		seed[i] = byte(i + 1)
	}
	priv := ed25519.NewKeyFromSeed(seed)
	pub := priv.Public().(ed25519.PublicKey)

	msg := []byte("hello")
	seedHex := hex.EncodeToString(seed)
	msgHex := hex.EncodeToString(msg)
	pubHex := hex.EncodeToString(pub)
	wantSig := hex.EncodeToString(ed25519.Sign(priv, msg))

	gotSig, err := bridge.Crypto.Sign(msgHex, seedHex)
	if err != nil {
		t.Fatalf("Crypto.Sign() failed: %v", err)
	}
	if gotSig != wantSig {
		t.Errorf("signature mismatch:\n got %s\nwant %s", gotSig, wantSig)
	}
	if !bridge.Crypto.Verify(gotSig, msgHex, pubHex) {
		t.Error("valid signature should verify against its public key")
	}
	// Wrong message must not verify.
	if bridge.Crypto.Verify(gotSig, hex.EncodeToString([]byte("goodbye")), pubHex) {
		t.Error("signature must not verify against a different message")
	}
}

// A fabricated signature must be rejected (mirrors Python test_crypto_verify_rejects_wrong_signature).
func TestCryptoVerifyRejectsWrongSignature(t *testing.T) {
	mnemonic, err := bridge.Crypto.GenerateMnemonic(24)
	if err != nil {
		t.Fatalf("GenerateMnemonic() failed: %v", err)
	}
	key, err := bridge.Crypto.DeriveKey(mnemonic, 0, 0, "payment")
	if err != nil {
		t.Fatalf("Crypto.DeriveKey() failed: %v", err)
	}
	pubKey := key.PublicKey
	fakeSig := strings.Repeat("00", 64)
	if bridge.Crypto.Verify(fakeSig, "68656c6c6f", pubKey) {
		t.Error("a fake all-zero signature must not verify")
	}
}

// --- Negative / Error Tests ---

func TestCryptoBlake2b256InvalidHex(t *testing.T) {
	_, err := bridge.Crypto.Blake2b256("not_valid_hex!")
	assertCclError(t, "Blake2b256(invalid hex)", err)
}

func TestCryptoSignInvalidKey(t *testing.T) {
	_, err := bridge.Crypto.Sign("68656c6c6f", strings.Repeat("zz", 32))
	assertCclError(t, "Sign(invalid key)", err)
}

File: wrappers/rust/tests/crypto_test.rs
//! Offline unit tests for the crypto namespace, ported to match the Python wrapper's coverage
//! (wrappers/python/tests/test_crypto.py). Covers exact Blake2b vectors, the 12-word mnemonic path,
//! and the negative / error cases (invalid hex, invalid signing key).

use mesmo::Bridge;

fn bridge() -> Bridge {
    Bridge::new().expect("Failed to create bridge")
}

#[test]
fn test_crypto_blake2b_256_known_vector() {
    // Blake2b-256 of "Hello" (0x48656c6c6f).
    let b = bridge();
    let hash = b.crypto().blake2b_256("48656c6c6f").expect("Failed to hash");
    assert_eq!(
        hash,
        "8b7ca7d27d9fc55fa30abfe515b3afb24e3fe89fdd02e2ac92bca2c96680642e"
    );
}

#[test]
fn test_crypto_blake2b_224_known_vector() {
    // Blake2b-224 of "Hello" (0x48656c6c6f).
    let b = bridge();
    let hash = b.crypto().blake2b_224("48656c6c6f").expect("Failed to hash");
    assert_eq!(
        hash,
        "376352b84882685054b2010033c2e9f479466fd10a609230240c06db"
    );
}

#[test]
fn test_crypto_generate_12_word_mnemonic() {
    let b = bridge();
    let mnemonic = b
        .crypto()
        .generate_mnemonic(12)
        .expect("Failed to generate mnemonic");
    assert_eq!(mnemonic.split_whitespace().count(), 12);
    assert!(b.crypto().validate_mnemonic(&mnemonic));
}

#[test]
fn test_crypto_invalid_mnemonic_rejected() {
    let b = bridge();
    assert!(!b.crypto().validate_mnemonic("not a valid mnemonic"));
}

// --- Negative / Error Tests ---

#[test]
fn test_crypto_blake2b_256_invalid_hex() {
    let b = bridge();
    let result = b.crypto().blake2b_256("not_valid_hex!");
    assert!(result.is_err(), "expected error for invalid hex input");
}

#[test]
fn test_crypto_sign_invalid_key() {
    let b = bridge();
    let bad_key = "zz".repeat(32);
    let result = b.crypto().sign("68656c6c6f", &bad_key);
    assert!(result.is_err(), "expected error signing with invalid key");
}
